### PR TITLE
Refactor rotest server to work with django api

### PR DESCRIPTION
**Pull request by IamShobe**

### CODE REVIEW

File: src/rotest/api/common/responses.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+class TokenResponseModel(AbstractResponse):
+    """Returns a token of the created sessions."""
+    TITLE = "Token Response"
+    PROPERTIES = [
+        StringField(name="token", description="Token of the current session",
+                    required=True)
+    ]
+
+
+class CleanupUserResponseModel(AbstractResponse):
```
Comment (osherdp) [inline src/rotest/api/common/responses.py]: I don't understand this response
In addition, the documentation doesn't seem related

File: src/rotest/api/common/responses.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    """Return details about the executed request."""
+    PROPERTIES = [
+        StringField(name="details", required=True)
+    ]
+
+
+class BadRequestResponseModel(AbstractResponse):
```
Comment (osherdp) [inline src/rotest/api/common/responses.py]: There's no special status code like 400?

File: src/rotest/api/common/responses.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class InfluencedResourcesResponseModel(AbstractResponse):
+    "Returns an array of the resources influenced by the action."
+    PROPERTIES = [
+        ArrayField(name="resource_descriptors", items_type=GenericModel,
+                   required=True)
+    ]
+
+
+class ShouldSkipResponse(AbstractResponse):
+    """Return if the test should skip and the reason why."""
```
Comment (osherdp) [inline src/rotest/api/common/responses.py]: Either ```Return``` of ```Returns``` here and everywhere. You can't have both

Comment (osherdp) [inline src/rotest/api/common/responses.py]: should be skipped

File: src/rotest/api/resource_control/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from cleanup_user import CleanupUser
```
Comment (osherdp) [inline src/rotest/api/resource_control/__init__.py]: Don't use relative imports
```from .cleanup_user import CleanupUser```

File: src/rotest/api/resource_control/cleanup_user.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    URI = "resources/cleanup_user"
+    DEFAULT_RESPONSES = {
+        httplib.OK: CleanupUserResponseModel,
+    }
+    TAGS = {
+        "post": ["Resources"]
+    }
+
+    def post(self, request, *args, **kwargs):
+        """Cleaning up user's requests and locked resources.
```
Comment (osherdp) [inline src/rotest/api/resource_control/cleanup_user.py]: *Clean

File: src/rotest/api/resource_control/cleanup_user.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        Args:
```
Comment (osherdp) [inline src/rotest/api/resource_control/cleanup_user.py]: I don't see this argument in the signature

File: src/rotest/api/resource_control/cleanup_user.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            SuccessReply. a reply indicating on a successful operation.
+        """
+        user_name = get_username(request)
+        with transaction.atomic():
+            resources = ResourceData.objects.select_for_update().filter(
+                owner=user_name)
+
+            if resources.count() == 0:
+                return Response({
+                    "details": "User {} didn't lock any resource".format(
```
Comment (osherdp) [inline src/rotest/api/resource_control/cleanup_user.py]: Why you're treating those cases differently?
The case where the user doesn't use any resource, is just a case where he locks 0 resources...

File: src/rotest/api/resource_control/lock_resources.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    def post(self, request, *args, **kwargs):
+        """Lock the given resources one by one.
+
+        Note:
+            If one of the resources fails to lock, all the resources that has
+            been locked until that resource will be released.
+        """
+        locked_resources = []
+        user_name = get_username(request)
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: username

File: src/rotest/api/resource_control/lock_resources.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            raise BadRequest("User {} has no matching object in the "
+                             "DB".format(user_name))
+
+        user = auth_models.User.objects.get(username=user_name)
+
+        groups = list(user.groups.all())
+        with transaction.atomic():
+            for descriptor_dict in descriptors:
+                try:
+                    desc = ResourceDescriptor.decode(descriptor_dict)
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: *descriptor

File: src/rotest/api/resource_control/lock_resources.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+                except Exception as e:
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: Any exception? That's very broad

File: src/rotest/api/resource_control/lock_resources.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                    raise BadRequest(e.message)
+                # query for resources that are usable and match the user's
+                # preference, which are either belong to a group he's in or
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: *groups

File: src/rotest/api/resource_control/lock_resources.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                        .filter(query).order_by('-reserved')
+
+                except FieldError as e:
+                    raise BadRequest(e.message)
+
+                if matches.count() == 0:
+                    raise BadRequest("No existing resource meets "
+                                     "the requirements: {!r}".format(desc))
+
+                availables = (resource for resource in matches
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: use ```first()```

File: src/rotest/api/resource_control/lock_resources.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    def _lock_resource(self, resource, user_name):
+        """Mark the resource as locked by the given user.
+
+        For complex resource, marks also its sub-resources as locked by the
+        given user.
+
+        Note:
+            The given resource *must* be available.
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: What does it mean that it must? If not, it's the programmer's mistake?

File: src/rotest/api/resource_control/lock_resources.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        user = auth_models.User.objects.get(username=user_name)
+
+        groups = list(user.groups.all())
+        with transaction.atomic():
+            for descriptor_dict in descriptors:
+                try:
+                    desc = ResourceDescriptor.decode(descriptor_dict)
+
+                except Exception as e:
+                    raise BadRequest(e.message)
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: append blank line

File: src/rotest/api/resource_control/lock_resources.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+            user_name (str): name of the locking user.
+        """
+        for sub_resource in resource.get_sub_resources():
+            self._lock_resource(sub_resource, user_name)
+
+        resource.owner = user_name
+        resource.owner_time = datetime.now()
+        resource.save()
+
+    def post(self, request, *args, **kwargs):
```
Comment (osherdp) [inline src/rotest/api/resource_control/lock_resources.py]: Way too long of a method. split

File: src/rotest/api/resource_control/query_resources.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        """Find and return the resources that answer the client's query.
+
+        Args:
+            request (Request): QueryResources request.
+
+        Returns:
+            ResourcesReply. a reply containing matching resources.
+        """
+        try:
+            desc = ResourceDescriptor.decode(request.model.obj)
```
Comment (osherdp) [inline src/rotest/api/resource_control/query_resources.py]: Here and everywhere, rename shortcuts

File: src/rotest/api/resource_control/query_resources.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+        except Exception as e:
+            raise BadRequest(e.message)
```
Comment (osherdp) [inline src/rotest/api/resource_control/query_resources.py]: You might encounter a bug of ours, which should return 500 instead...

File: src/rotest/api/resource_control/release_resources.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    def _release_resource(self, resource, user_name):
+        """Mark the resource as free.
+
+        For complex resource, marks also its sub-resources as free.
+
+        Args:
+            resource (ResourceData): resource to release.
+            user_name (str): name of the releasing user.
```
Comment (osherdp) [inline src/rotest/api/resource_control/release_resources.py]: Everywhere ```username```

File: src/rotest/api/resource_control/release_resources.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+            ResourceAlreadyAvailableError: if resource was already available.
+        """
+        errors = {}
+
+        for sub_resource in resource.get_sub_resources():
+            try:
+                self._release_resource(sub_resource, user_name)
+
+            except ServerError as ex:
+                errors[sub_resource.name] = (ex.ERROR_CODE, str(ex))
```
Comment (osherdp) [inline src/rotest/api/resource_control/release_resources.py]: I think you're not including all of the traceback
Make sure

File: src/rotest/api/resource_control/update_fields.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    URI = "resources/update_fields"
+    DEFAULT_MODEL = UpdateFieldsParamsModel
+    DEFAULT_RESPONSES = {
+        httplib.NO_CONTENT: SuccessResponse,
+    }
+    TAGS = {
+        "post": ["Resources"]
+    }
+
+    def post(self, request, *args, **kwargs):
```
Comment (osherdp) [inline src/rotest/api/resource_control/update_fields.py]: It's actually should be a ```PUT``` request

File: src/rotest/api/test_control/middleware.py [hunk 21]
```diff
@@ -0,0 +1,26 @@
+"""Contain middleware and session handling of test control views."""
+# pylint: disable=unused-argument, no-self-use
+SESSIONS = {}
+
+
+def session_middleware(get_response):
+    """Inject sessions to all test control views."""
```
Comment (osherdp) [inline src/rotest/api/test_control/middleware.py]: doc
what is this ```get_response```?

File: src/rotest/api/common/responses.py (hunk 2 above)
Comment (UnDarkle) [reply src/rotest/api/common/responses.py]: I think BadRequest is not right here.
we send it when the resources are locked, for example.
maybe FailureResponseModel would be better?

File: src/rotest/management/client/manager.py [hunk 22]
```diff
@@ -276,14 +291,28 @@ def _lock_resources(self, descriptors, timeout=None):
...
-                             zip(server_requests, reply.resources))
+            request_data = LockResourcesParamsModel({
+                "descriptors": encoded_requests
+            })
+            start_time = time.time()
+            while True:
+                response = self.requester.request(LockResources,
+                                                  data=request_data,
+                                                  method="post")
+                if isinstance(response, BadRequestResponseModel):
```
Comment (UnDarkle) [inline src/rotest/management/client/manager.py]: shouldn't you check the type of the response?
like, if the user is not recognized retrying won't help....

File: tests/core/multiprocess/test_runner.py [hunk 23]
```diff
@@ -34,6 +37,8 @@ class AbstractMultiprocessRunnerTest(BasicRotestUnitTest):
 
     fixtures = ['case_ut.json']
 
+    @mock.patch("rotest.management.client.client.Requester",
```
Comment (UnDarkle) [inline tests/core/multiprocess/test_runner.py]: found out why you need this...
the clients in the worker she be mock clients

File: tests/management/test_resource_manager.py [hunk 24]
```diff
@@ -1002,7 +1004,7 @@ def test_update_fields(self):
                           (resource2.version, 2))
 
         # Change version of a single resource.
-        self.client.update_fields(DemoResourceData,
+        self.client.update_fields(DemoResource,
```
Comment (UnDarkle) [inline tests/management/test_resource_manager.py]: what? that should be the model, not the resource class!

File: setup.py [hunk 25]
```diff
@@ -40,7 +40,8 @@
     'colorama',
     'termcolor',
     'jsonschema',
-    'basicstruct'
+    'basicstruct',
+    'swaggapi==0.5.0'
```
[the remark this answers is not in the document]
Comment (IamShobe) [reply setup.py]: what if the api of the library breaks?

File: src/rotest/api/common/models.py [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+    """Generic model can be any dict object."""
+    TITLE = "Generic Object"
+    PROPERTIES = []
+    EXAMPLE = {}
+
+
+class ResourceDescriptorModel(AbstractAPIModel):
+    """Descriptor of a resource.
+
+    Args:
```
[the remark this answers is not in the document]
Comment (IamShobe) [reply src/rotest/api/common/models.py]: Actually I meant Args - 
Models are actually kind of functions if u think about it.
I want to doc what are each of the properties do.
The class Attributes are ir-relevant

File: src/rotest/api/common/models.py [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+        class (str): the classname of the test.
+        subtests (list): list of TestModel. the subtests of the current test.
+    """
+    TITLE = "Test"
+    PROPERTIES = [
+        NumberField(name="id"),
+        StringField(name="name"),
+        StringField(name="class"),
+        ArrayField(name="subtests",
+                   items_type=DynamicType("TestModel",
```
[the remark this answers is not in the document]
Comment (IamShobe) [reply src/rotest/api/common/models.py]: It is impassible in the launguge to use a referance of something before declaring it.
Here the class is on the declaration state, therefore it cannot find itself - because it is yet to be declared.
To solve that, there are 2 ways to approch it -
 1- to create a new type which dynamically imports the model
 2- using eval("TestModel")

I think 1 is a better option - as a closure.

File: src/rotest/api/common/responses.py (hunk 2 above)
Comment (IamShobe) [reply src/rotest/api/common/responses.py]: @osherdp  the status code is irrelavant to the content sent to the user, therefore we  need to specify the structure of the response.
@UnDarkle I can relate to that.

File: src/rotest/api/resource_control/cleanup_user.py (hunk 6 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/cleanup_user.py]: Because  it is a new  concept - the models, I wanted to describe what the function  model should actually receive.
We need to  decide what  is the convention  for that

File: src/rotest/api/resource_control/cleanup_user.py (hunk 7 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/cleanup_user.py]: This is copy pasted from the original code.
I am not sure what u  meant here.
U want me to return "User {} was successfully cleaned always"?

File: src/rotest/api/resource_control/lock_resources.py (hunk 13 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/lock_resources.py]: This is copy pasted from the original code.
I dont know. maybe @UnDarkle knows?
maybe it messes up the user cleanup.. i am not sure

File: src/rotest/api/resource_control/query_resources.py (hunk 17 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/query_resources.py]: What bug?

File: src/rotest/api/resource_control/lock_resources.py (hunk 12 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/lock_resources.py]: doesn't match my purpose

File: src/rotest/api/resource_control/release_resources.py (hunk 19 above)
Comment (IamShobe) [reply src/rotest/api/resource_control/release_resources.py]: i am not sure what u meant?
the exception except statement?
or the error caught?
both of them are a copy paste  from the original code..
i am not sure what to do then.

File: src/rotest/management/client/result_client.py [hunk 28]
```diff
@@ -60,45 +77,78 @@ def start_test_run(self, main_test):
...
-    def update_run_data(self, run_data):
+        request_data = StartTestRunParamsModel({
+            "tests": tests_tree_dict,
+            "run_data": run_data
+        })
+        response = self.requester.request(StartTestRun,
+                                          data=request_data,
+                                          method="post")
+        if isinstance(response, BadRequestResponseModel):
+            raise RuntimeError(response.details)
```
Comment (osherdp) [inline src/rotest/management/client/result_client.py]: Above it was ```Exception```. Why?

File: src/rotest/management/common/json_parser.py [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+from rotest.management.base_resource import BaseResource
+from rotest.management.common.parsers.abstract_parser import ParsingError
+from rotest.management.common.utils import (TYPE_NAME, DATA_NAME, PROPERTIES,
+                                            extract_type, extract_type_path)
+
+
+class JSONParser(object):
+    """json messages parser.
+
+    This message parser allow to encode & decode resource management messages.
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: *allows

File: src/rotest/management/common/json_parser.py [hunk 30]
```diff
@@ -0,0 +1,378 @@
...
+    and booleans), lists, dictionaries, and resources (which derived from
+    :class:`BaseResource`).
+
+    Any other object type may cause an error during the encoding process.
+    The result of successful encoding process is an json string, represent the
+    encoded message.
+
+    For instance, the inner message object:
+        [False, {'key1': 5, 'key2': 'google'}, [1, 2, 3]]
+    will be encoded as:
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: ?

File: tests/core/multiprocess/test_runner.py (hunk 23 above)
Comment (IamShobe) [reply tests/core/multiprocess/test_runner.py]: ? how

File: src/rotest/management/common/json_parser.py [hunk 31]
```diff
@@ -0,0 +1,378 @@
...
+            resource_data (ResourceData): a resource to encode.
+
+        Returns:
+            object. encoded data, depends on the parser type.
+
+        Raises:
+            TypeError: given 'resource' is not of type 'ResourceData'.
+            ParsingError: encoding failure.
+        """
+
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: Here and everywhere: delete

File: src/rotest/management/common/json_parser.py [hunk 32]
```diff
@@ -0,0 +1,378 @@
...
+            data (object): data to decode, encoded data of 'AbstractMessage',
+                depends on the parser type.
+
+        Returns:
+            AbstractMessage. decoded message.
+
+        Raises:
+            ParsingError: decoding failure.
+        """
+        # try:
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: ?

File: src/rotest/management/common/json_parser.py [hunk 33]
```diff
@@ -0,0 +1,378 @@
...
+        # except Exception as err:
+        #     raise ParsingError("Decoding data %r has failed. Reason: %s." %
+        #                        (data, err))
+
+    def _encode(self, data):
+        """Encode the given data according to its type.
+
+        Warning:
+            Do not change the order of the type validation! Some types matches
+            more than one type(e.g: bool is also a Number).
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: add space before (

File: src/rotest/management/common/json_parser.py [hunk 34]
```diff
@@ -0,0 +1,378 @@
...
+        Args:
+            data (object): an object to encode.
+
+        Returns:
+            object. encoded data, depends on the parser type.
+
+        Raises:
+            TypeError: given 'data' couldn't be encoded by this parser.
+        """
+        if data is None:
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: Replace this whole method with a dictionary

File: src/rotest/management/common/json_parser.py [hunk 35]
```diff
@@ -0,0 +1,378 @@
...
+            return data
+
+        if isinstance(data, Number):
+            return data
+
+        if isinstance(data, dict):
+            return self._encode_dict(data)
+
+        if isinstance(data, (list, tuple)):
+            return self._encode_list(data)
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: We really can't just use ```json``` for this whole functionality?

File: src/rotest/management/common/utils.py [hunk 36]
```diff
@@ -23,6 +22,18 @@
 LOCALHOST = "localhost"
 
 
+def get_client_ip(request):
+    """Get client's ip address."""
```
Comment (osherdp) [inline src/rotest/management/common/utils.py]: Returns section...

File: src/rotest/management/common/utils.py [hunk 37]
```diff
@@ -40,6 +51,11 @@ def get_host_name(ip_address):
     return host
 
 
+def get_username(request):
+    """Return the username."""
```
Comment (osherdp) [inline src/rotest/management/common/utils.py]: same

File: tests/management/test_parallel.py [hunk 38]
```diff
@@ -1,263 +0,0 @@
-"""Tests resource manager ability to work with multiple clients in parallel.
```
Comment (osherdp) [inline tests/management/test_parallel.py]: Why you removed the whole file?

File: src/rotest/management/common/json_parser.py (hunk 35 above)
Comment (IamShobe) [reply src/rotest/management/common/json_parser.py]: i wanted it to be as  close as  possible to the original XML parser so i wouldn't have to do much changes.
I think it should be another  issue if  u  want  to refactored that.
It creates a complex structure and I  am  not sure  if it will break something.

File: src/rotest/api/test_control/should_skip.py [hunk 39]
```diff
@@ -0,0 +1,52 @@
...
+    DEFAULT_MODEL = TestControlOperationParamsModel
+    DEFAULT_RESPONSES = {
+        httplib.OK: ShouldSkipResponse,
+    }
+    TAGS = {
+        "post": ["Tests"]
+    }
+
+    @session_middleware
+    def post(self, request, sessions, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (IamShobe) [reply src/rotest/api/test_control/should_skip.py]: you want me to pass the  parameters in the url?

[the remark this answers is not in the document]
Comment (IamShobe) [reply src/rotest/api/test_control/should_skip.py]: i think  it doesnt make sense too.. (to make  it a get request).
I wanted to pass the parameters in the body.

File: src/rotest/management/common/json_parser.py [hunk 40]
```diff
@@ -113,34 +121,17 @@ def _encode(self, data):
...
-        if isinstance(data, (list, tuple)):
-            return self._encode_list(data)
-
-        if isinstance(data, type):
-            return self._encode_class(data)
+        base_types = [basestring, bool, Number]
 
-        if isinstance(data, models.Model):
-            return self._encode_resource_data(data)
+        for encoder_type in base_types:
```
Comment (osherdp) [inline src/rotest/management/common/json_parser.py]: no more for.
use instead:
data_type = type(data)
dict[data_type] or dict.get(data_type) etc

File: src/rotest/management/common/utils.py [hunk 41]
```diff
@@ -23,7 +23,11 @@
 
 
 def get_client_ip(request):
-    """Get client's ip address."""
+    """Get client's ip address.
```
Comment (osherdp) [inline src/rotest/management/common/utils.py]: not describing what the func does =\ explain better

File: setup.py [hunk 42]
```diff
@@ -40,7 +40,8 @@
     'colorama',
     'termcolor',
     'jsonschema',
-    'basicstruct'
+    'basicstruct',
+    'swaggapi'
```
Comment (osherdp) [inline setup.py]: requests is not a direct dependency?

File: src/rotest/api/resource_control/lock_resources.py (hunk 13 above)
Comment (osherdp) [reply src/rotest/api/resource_control/lock_resources.py]: Fix it

File: src/rotest/api/common/responses.py (hunk 1 above)
Comment (osherdp) [reply src/rotest/api/common/responses.py]: ?

File: src/rotest/api/resource_control/cleanup_user.py (hunk 7 above)
Comment (osherdp) [reply src/rotest/api/resource_control/cleanup_user.py]: I want lines 37-41 deleted

File: tests/core/multiprocess/test_runner.py (hunk 23 above)
Comment (IamShobe) [reply tests/core/multiprocess/test_runner.py]: solved together

File: docs/configurations.rst [hunk 43]
```diff
@@ -103,14 +103,14 @@ that the client should contact with, use the following methods:
 Port
 ----
 
-.. envvar:: ROTEST_SERVER_PORT
+.. envvar:: DJANGO_SERVER_PORT
```
Comment (osherdp) [inline docs/configurations.rst]: prefix of ROTEST

File: docs/intro/basic_usage.rst [hunk 44]
```diff
@@ -92,6 +111,8 @@ Pay attention to the following:
 * The `host` key is how the client should contact the server. It's an IP
   address, or a DNS of the server. For now, both the client and server are
   running on the same machine., but it doesn't have to be that way.
+* The `api_base_url` key is the end-point url of the rotest urls file,
+  it must be configured for the resource manager to work - default is "rotest/api/".
```
Comment (osherdp) [inline docs/intro/basic_usage.rst]: resource management

File: src/rotest/api/common/responses.py (hunk 1 above)
Comment (osherdp) [reply src/rotest/api/common/responses.py]: delete

File: src/rotest/management/client/manager.py [hunk 45]
```diff
@@ -276,14 +295,33 @@ def _lock_resources(self, descriptors, timeout=None):
...
+                "descriptors": encoded_requests
+            })
+            start_time = time.time()
+            while True:
+                response = self.requester.request(LockResources,
+                                                  data=request_data,
+                                                  method="post")
+                if isinstance(response, FailureResponseModel):
+                    match = re.match(USER_NOT_EXIST.format(".*"),
+                                 response.details)
```
Comment (osherdp) [inline src/rotest/management/client/manager.py]: indent

File: src/rotest/api/common/models.py [hunk 46]
```diff
@@ -0,0 +1,189 @@
...
+        ArrayField(name="descriptors", items_type=ResourceDescriptorModel,
+                   required=True)
+    ]
+
+
+class ReleaseResourcesParamsModel(AbstractAPIModel):
+    """Release the given resources names.
+
+    Args:
+        resources (list): list of str. the names of the resources to be
```
Comment (osherdp) [inline src/rotest/api/common/models.py]: resource names to be released
